### PR TITLE
storage: Use mock values during some pixel tests

### PR DESCRIPTION
**Pull request by mvollmer**
Instead of ignoring parts of the DOM that are unpredictable, we
replace them with some constant non-sense.  This gets rid of the
unpredictable content and also produces a predictable layout by
avoiding variable sized content.

### CODE REVIEW

**Comment by mvollmer** [comment]
We don't need to restore the old values; the idea is that React will do that on the next render.  But there is a race between setting the mock text content and taking the snapshot.  If React decides to render in that window, things will break. Let's see.

**Comment by garrett** [comment]
> We don't need to restore the old values; the idea is that React will do that on the next render. But there is a race between setting the mock text content and taking the snapshot. If React decides to render in that window, things will break. Let's see.

No, I don't think this is the right solution, as we're still setting up a race condition.

The data itself has to change before the DOM.

**Comment by mvollmer** [comment]
> No, I don't think this is the right solution, as we're still setting up a race condition.

We could have a global toggle that puts each page into "mock mode" where it will replace certain elements with fake data.  This means changes to our code, but that's totally acceptable.

Or, we could try to stop the pages JavaScript via the CDP Debugger.pause() function... but maybe that will also block our own eval_js and we need to use CDP also to poke in the fake data... sounds a bit too involved to me.

**Comment by garrett** [comment]
> We could have a global toggle that puts each page into "mock mode" where it will replace certain elements with fake data. This means changes to our code, but that's totally acceptable.

I was wondering about this and was going to suggest it, but the ramifications could be immense, as so much can change everywhere.

> Or, we could try to stop the pages JavaScript via the CDP Debugger.pause() function... but maybe that will also block our own eval_js and we need to use CDP also to poke in the fake data... sounds a bit too involved to me.

I have been looking for ways to pause React's DOM updating instead. The problem isn't with browser DOM updates, but with React updating the DOM, so we might as well try to attack it at that level.

If we _can_ do something like `React.RenderUpdates.Pause()` (this is totally made up), swap out the string(s) in the DOM, then test, then `React.RenderUpdates.Resume()` (also made up), then this simple string swap approach would work fine.

Surely, we're not the first to hit this kind of problem. Right?

**Comment by garrett** [comment]
`shouldComponentUpdate()` might almost be what we want.

https://reactjs.org/docs/react-component.html#shouldcomponentupdate

Use shouldComponentUpdate() to let React know if a component’s output is not affected by the current change in state or props.

_HOWEVER:_ 

> This method only exists as a [performance optimization](https://reactjs.org/docs/optimizing-performance.html). Do not rely on it to “prevent” a rendering, as this can lead to bugs.

and

> In the future React may treat shouldComponentUpdate() as a hint rather than a strict directive, and returning false may still result in a re-rendering of the component.

So I guess we should keep looking.

It would also be nice to flip the theoretical pause switch on a React level, not on a component level.

**Comment by garrett** [comment]
There's `ReactDOM.flushSync()` (https://reactjs.org/docs/react-dom.html#flushsync) which causes React to render immediately. If we called that in the test code (_important: not in production code_) before monkeying with the DOM, then it's a way to prevent some of the race conditions. It's a bad hack though; it would be better to pase/resume React's rendering somehow.

**Comment by mvollmer** [comment]
`Debugger.pause` works well enough, but changing the DOM via the CDP API is cumbersome and apparently also unreliable.  Hmm.

**Comment by mvollmer** [comment]
> and apparently also unreliable

No, i's just that querySelectorAll doesn't understand the`:contains` syntax.

**Comment by mvollmer** [comment]
> I have been looking for ways to pause React's DOM updating instead.

Yeah, me too, but I didn't find anything.

React renders should not be totally random, there needs to be an event of some kind. During testing, we control all input events.  That leaves the WebSocket events.  (Idle and timeout events hopefully don't matter...)

**Comment by mvollmer** [comment]
So I am reasonably happy with the Debugger.pause approach.  Code could be cleaned up, of course.

**Comment by mvollmer** [comment]
Arg, now I get this:

```
RuntimeError: DOMException: Failed to execute 'removeChild' on 'Node': The node to be removed is not a child of this node.
```

This originates from React, so I am guessing React is unhappy with our changes to the DOM when it tries to re-render.  This would be a show-stopper for the "mock without React involvement" approach.

**Comment by garrett** [comment]
> React renders should not be totally random, there needs to be an event of some kind. During testing, we control all input events. That leaves the WebSocket events. (Idle and timeout events hopefully don't matter...)

No, afaik, React updates the DOM during render calls, and there could be a different render call on the page which could trigger it being updated. It's not necessarily tied to input events.

Again: if we cannot fix the data (static IP), then having mock data is probably the more correct approach here. Otherwise we'll constantly be fighting React and having race conditions.

**Comment by mvollmer** [comment]
Alright, what about the global "mock mode", where the tests instruct the page to behave differenty just for the benefits of the tests.  I think that is a useful idea in general, and allows us to test things that are otherwise hard to set up "for real".

I have pushed a minimal version of this idea that just replaces certain texts, via a new hook.

**Comment by garrett** [comment]
> Alright, what about the global "mock mode", where the tests instruct the page to behave differenty just for the benefits of the tests. I think that is a useful idea in general, and allows us to test things that are otherwise hard to set up "for real".

Yes, that makes sense.

The big issue with the IPv6 addresses is that we need to be able to swap out data, not just add it, so this should be taken into consideration.

**Comment by garrett** [comment]
Alternative is to be able to not show stuff when in mock mode, so things can be hidden and alternatives can be shown, instead of changing it.

File: pkg/lib/hooks.js [hunk 1]
```diff
@@ -325,3 +325,25 @@ export function useEvent(obj, event, handler) {
...
+    let config = useMockConfig();
+
+    return (val, tag) => {
+        if (!config)
+            return val;
+        if (topic)
+            config = config[topic];
+        if (tag && config)
+            config = config[tag];
+        return config || "mock";
```
Comment (cockpituous) [inline pkg/lib/hooks.js:347]: These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220819-071652-20323b89-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/network-main.jsx [hunk 2]
```diff
@@ -92,8 +93,8 @@ export const NetworkPage = ({ privileged, operationInProgress, usage_monitor, pl
 
         if (show_traffic) {
             const samples = usage_monitor.samples[iface.Name];
-            row.columns.push({ title: samples ? cockpit.format_bits_per_sec(samples[1][0] * 8) : "" });
-            row.columns.push({ title: samples ? cockpit.format_bits_per_sec(samples[0][0] * 8) : "" });
+            row.columns.push({ title: samples ? mock(cockpit.format_bits_per_sec(samples[1][0] * 8), "tx") : "" });
+            row.columns.push({ title: samples ? mock(cockpit.format_bits_per_sec(samples[0][0] * 8), "rx") : "" });
```
Comment (cockpituous) [inline pkg/networkmanager/network-main.jsx:97]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220819-071652-20323b89-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/hooks.js [hunk 3]
```diff
@@ -325,3 +325,29 @@ export function useEvent(obj, event, handler) {
...
+    return (val, tag) => {
+        let mock = config;
+
+        if (!mock)
+            return val;
+
+        if (topic)
+            mock = mock[topic];
+        if (tag && mock)
+            mock = mock[tag];
```
Comment (cockpituous) [inline pkg/lib/hooks.js:349]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220819-083129-c0e512fa-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/hooks.js [hunk 4]
```diff
@@ -325,3 +325,29 @@
...
+
+        return mock || "mock";
```
Comment (cockpituous) [inline pkg/lib/hooks.js:351]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220819-083129-c0e512fa-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/network-main.jsx (hunk 2 above)
Comment (cockpituous) [inline pkg/networkmanager/network-main.jsx:97]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220819-083129-c0e512fa-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/hooks.js [hunk 5]
```diff
@@ -325,3 +325,21 @@ export function useEvent(obj, event, handler) {
...
+    useEvent(window, "test_mock_changed");
+    return window.test_mock;
+}
+
+export function useMock(topic) {
+    const config = useMockConfig();
+
+    return (val, mock_val) => {
+        if (config === true || (topic && config === topic))
+            return mock_val || "mock";
```
Comment (cockpituous) [inline pkg/lib/hooks.js:342]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220822-064711-e2f09928-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/network-main.jsx [hunk 6]
```diff
@@ -92,8 +93,8 @@ export const NetworkPage = ({ privileged, operationInProgress, usage_monitor, pl
 
         if (show_traffic) {
             const samples = usage_monitor.samples[iface.Name];
-            row.columns.push({ title: samples ? cockpit.format_bits_per_sec(samples[1][0] * 8) : "" });
-            row.columns.push({ title: samples ? cockpit.format_bits_per_sec(samples[0][0] * 8) : "" });
+            row.columns.push({ title: samples ? mock(cockpit.format_bits_per_sec(samples[1][0] * 8)) : "" });
+            row.columns.push({ title: samples ? mock(cockpit.format_bits_per_sec(samples[0][0] * 8)) : "" });
```
Comment (cockpituous) [inline pkg/networkmanager/network-main.jsx:97]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220822-064711-e2f09928-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/hooks.js (hunk 5 above)
Comment (cockpituous) [inline pkg/lib/hooks.js:342]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220822-092047-5ac70a0f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/network-main.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/networkmanager/network-main.jsx:97]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220822-092047-5ac70a0f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: test/verify/check-storage-mounting [hunk 7]
```diff
@@ -72,11 +72,8 @@ ExecStart=/usr/bin/sleep infinity
 
         self.content_dropdown_action(1, "Unmount")
         b.wait_in_text("#dialog", str(sleep_pid))
-        b.wait_in_text("#dialog", "sleep infinity")
```
Comment (mvollmer) [inline test/verify/check-storage-mounting:75]: These assertions should not be removed.

File: pkg/lib/hooks.js (hunk 5 above)
Comment (cockpituous) [inline pkg/lib/hooks.js:342]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220823-070904-893c0860-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/networkmanager/network-main.jsx (hunk 6 above)
Comment (cockpituous) [inline pkg/networkmanager/network-main.jsx:97]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220823-070904-893c0860-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/hooks.js (hunk 5 above)
Comment (cockpituous) [inline pkg/lib/hooks.js:342]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220823-091359-727e4c6c-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

Comment (cockpituous) [inline pkg/lib/hooks.js:342]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220823-130157-33c3a71e-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by mvollmer** [comment]
In defense of the general idea of putting fake data into a page while doing pixel tests:

We already have a mechanism to ignore certain DOM elements.  This is necessary if there are any elements that have unpredictable and hard to control content, such as UUIDs of newly created things. However, if the unpredictable content varies enough, it will influence the layout of elements around it.  Sometimes this influence is a bug; no matter what an element shows, we don't want it to change the layout.  Sometimes, it is not a bug, such as with table columns.

The way tables work with Patternfly is that even the smallest change in one column (such as replacing a "9" with a narrower "1") will change the position of all columns, also by a very small amount.  This will shift all texts by half a pixel, say, and will look like unreliable text rendering.

In those cases, it helps to make sure that all columns always have a constant, predictable content.

In the case where a layout change counts as a bug, we can use the mocking mechanism to test extreme cases that might be hard to produce in other ways.

File: pkg/lib/hooks.js (hunk 5 above)
Comment (cockpituous) [inline pkg/lib/hooks.js:342]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220824-103107-617c38e7-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by mvollmer** [comment]
Pixel tests are supposed to fail for the time being.  Let's fix them immediately before merging.

File: pkg/lib/hooks.js (hunk 5 above)
Comment (cockpituous) [inline pkg/lib/hooks.js:342]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20220927-081450-953815f3-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/lib/hooks.js [hunk 8]
```diff
@@ -326,3 +326,21 @@ export function useEvent(obj, event, handler) {
...
```
Comment (cockpituous) [inline pkg/lib/hooks.js:343]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230302-084754-aa03fba9-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 9]
```diff
@@ -1161,13 +1163,15 @@ export function init_active_usage_processes(client, usage) {
 }
 
 export const StopProcessesMessage = ({ mount_point, users }) => {
+    const mock = useMock()?.busy;
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1166]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230302-101354-40692ee9-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 10]
```diff
@@ -1161,13 +1163,15 @@
...
+
     const process_rows = users.filter(u => u.pid).map(u => {
         return {
             columns: [
-                u.pid,
+                mock?.pid ?? u.pid,
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1171]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230302-101354-40692ee9-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 11]
```diff
@@ -1161,13 +1163,15 @@
...
                 { title: u.cmd.substr(0, 100), props: { modifier: "breakWord" } },
                 u.user || "-",
-                { title: format_delay(-u.since * 1000), props: { modifier: "nowrap" } }
+                { title: mock?.since ?? format_delay(-u.since * 1000), props: { modifier: "nowrap" } }
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1174]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230302-101354-40692ee9-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/part-tab.jsx [hunk 12]
```diff
@@ -22,35 +22,35 @@ import { DescriptionList, DescriptionListDescription, DescriptionListGroup, Desc
...
-                <DescriptionListGroup>
-                    <DescriptionListTerm>{_("Type")}</DescriptionListTerm>
-                    <DescriptionListDescription>{block_part.Type}</DescriptionListDescription>
-                </DescriptionListGroup>
-            </DescriptionList>
-        );
-    }
-}
+export const PartitionTab = ({ client, block }) => {
+    const mock_UUID = useMock()?.part_UUID;
```
Comment (cockpituous) [inline pkg/storaged/part-tab.jsx:30]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230302-101354-40692ee9-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx (hunk 9 above)
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1166]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230302-122128-2c63e908-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx (hunk 10 above)
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1171]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230302-122128-2c63e908-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx (hunk 11 above)
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1174]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230302-122128-2c63e908-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/part-tab.jsx (hunk 12 above)
Comment (cockpituous) [inline pkg/storaged/part-tab.jsx:30]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230302-122128-2c63e908-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 13]
```diff
@@ -1068,6 +1070,7 @@ export const BlockingMessage = (usage) => {
 };
 
 const UsersPopover = ({ users }) => {
+    const mock = useMock()?.busy;
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1073]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230322-115306-eab6c63e-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 14]
```diff
@@ -1094,7 +1097,7 @@
                         ? <p>
                             <b>{_("Processes using the location")}</b>
                             <List>
-                                { processes.slice(0, max).map((u, i) => <ListItem key={i}>{u.comm} (user: {u.user}, pid: {u.pid})</ListItem>) }
+                                { processes.slice(0, max).map((u, i) => <ListItem key={i}>{u.comm} (user: {u.user}, pid: {mock?.pid ?? u.pid})</ListItem>) }
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1100]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230322-115306-eab6c63e-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 15]
```diff
@@ -1175,13 +1178,15 @@
 }
 
 export const StopProcessesMessage = ({ mount_point, users }) => {
+    const mock = useMock()?.busy;
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1181]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230322-115306-eab6c63e-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 16]
```diff
@@ -1175,13 +1178,15 @@
...
+
     const process_rows = users.filter(u => u.pid).map(u => {
         return {
             columns: [
-                u.pid,
+                mock?.pid ?? u.pid,
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1186]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230322-115306-eab6c63e-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/dialog.jsx [hunk 17]
```diff
@@ -1175,13 +1178,15 @@
...
                 { title: u.cmd.substr(0, 100), props: { modifier: "breakWord" } },
                 u.user || "-",
-                { title: format_delay(-u.since * 1000), props: { modifier: "nowrap" } }
+                { title: mock?.since ?? format_delay(-u.since * 1000), props: { modifier: "nowrap" } }
```
Comment (cockpituous) [inline pkg/storaged/dialog.jsx:1189]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230322-115306-eab6c63e-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/part-tab.jsx (hunk 12 above)
Comment (cockpituous) [inline pkg/storaged/part-tab.jsx:30]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17654-20230322-115306-eab6c63e-fedora-37-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by mvollmer** [comment]
Alright, nobody likes to change the production code, including me and the Coverage report.

Summary of past approaches:

 - Changing the DOM without Reacts involvement might be racy, but we can get that done with CDP and Debugger.pause.
  - Forcefully changing the DOM and then hoping that React puts its back how it was doesn't work because React will stop working when it finds the DOM in an unexpected state.
 
So, I'll try to modify the DOM with Debugger.pause and put it back how it was before unfreezing the page. To make this simply, I'll try to get the desired mocking effect by just adding things that can be cleanly taken away again, like `display: none`, and using `:after` .

**Comment by mvollmer** [comment]
> This means changes to our code, but that's totally acceptable.

Nope, that wasn't acceptable. :-)

**Comment by KKoukiou** [comment]
@mvollmer shall we close this one then as the implementation with the dom manipulation from testlib was prefered?

**Comment by mvollmer** [comment]
> @mvollmer shall we close this one then as the implementation with the dom manipulation from testlib was prefered?

Yes, let's.

New story in #18772.

**Pull request closed without merging**